internal/server/create_document.go

```diff
@@ -0,0 +1,293 @@
+					Postcode:              postFormString(r, "postcode"),
+					IsAirmailRequired:     postFormString(r, "isAirmailRequired") == "Yes",
+					PhoneNumber:           postFormString(r, "phoneNumber"),
+					Email:                 postFormString(r, "email"),
+					CorrespondenceByPost:  postFormCheckboxChecked(r, "correspondenceBy", "post"),
+					CorrespondenceByEmail: postFormCheckboxChecked(r, "correspondenceBy", "email"),
+					CorrespondenceByPhone: postFormCheckboxChecked(r, "correspondenceBy", "phone"),
+					CorrespondenceByWelsh: postFormCheckboxChecked(r, "correspondenceBy", "welsh"),
+				}
+
+				createdContact, err := client.CreateContact(ctx, contact)
+
+				if ve, ok := err.(sirius.ValidationError); ok {
+					w.WriteHeader(http.StatusBadRequest)
+					data.Error = ve
+					data.Recipients = append(data.Recipients, contact)
+				} else if err != nil {
+					return err
+				} else {
+					data.Success = true
+					data.TemplateSelected.TemplateId = r.FormValue("templateId")
+					data.SelectedInserts = r.Form["insert"]
+					data.HasViewedInsertPage = true
+					data.Recipients, err = getRecipients(ctx, client, data.Case)
+					if err != nil {
+						return err
+					}
+					data.Recipients = append(data.Recipients, createdContact)
+				}
+			}
+		}
+
+		return tmpl(w, data)
+	}
+}
+
+func getRecipients(ctx sirius.Context, client CreateDocumentClient, caseItem sirius.Case) ([]sirius.Person, error) {
+	var recipientIds []int
+	donor := *caseItem.Donor
+	recipientIds = append(recipientIds, donor.ID)
+	recipientIds = append(recipientIds, getPersonIds(caseItem.TrustCorporations)...)
+	recipientIds = append(recipientIds, getPersonIds(caseItem.Attorneys)...)
+
+	var recipients []sirius.Person
+	group, groupCtx := errgroup.WithContext(ctx.Context)
+	var personsMu sync.Mutex
+
+	for _, recipientID := range recipientIds {
+		recipientID := recipientID
+
+		group.Go(func() error {
+			person, err := client.Person(ctx.With(groupCtx), recipientID)
+			if err != nil {
+				return err
+			}
+
+			personsMu.Lock()
+			recipients = append(recipients, person)
+			personsMu.Unlock()
+			return nil
+		})
+	}
+
+	if err := group.Wait(); err != nil {
+		return nil, err
+	}
+
+	return recipients, nil
+}
+
+func getPersonIds(persons []sirius.Person) []int {
+	var personIds []int
+	for _, person := range persons {
+		personIds = append(personIds, person.ID)
+	}
+	return personIds
+}
+
+func translateDocumentData(documentTemplateData []sirius.DocumentTemplateData, documentTemplateRefData []sirius.RefDataItem) []sirius.RefDataItem {
+	var documentTemplateTypes []sirius.RefDataItem
+	for _, dt := range documentTemplateData {
+		for _, refData := range documentTemplateRefData {
+			if refData.Handle == dt.OnScreenSummary {
+				translatedRefDataItem := sirius.RefDataItem{
+					Handle:         dt.TemplateId,
+					Label:          refData.Label,
+					UserSelectable: false,
+				}
+				documentTemplateTypes = append(documentTemplateTypes, translatedRefDataItem)
```

Comment:
could `break` after this to save going through the whole `documentTemplateRefData` list